### ISSUE TITLE
Is it possible to remove the dependency on vc redistributable and .NET framework from the installer?

### ISSUE THREAD

**Issue by lins05** [dokan-dev/dokany#710]
First, thanks for the great work of the dokan project!

The [seafile drive client](https://github.com/haiwen/seafile-user-manual/blob/master/en/drive_client/using_drive_client.md) installer relies on the dokan installer. And currently one frequent issue reported by the users is that sometimes the install process failed during the installation of the vc redistributable. So we're checking the possibility of removing the dep on vc redist.

My understanding is that the dokan installer requires the vc redist because there are several custom actions defined with C#, in [dokan_wix/WiXCA/CustomAction.cs](https://github.com/dokan-dev/dokany/blob/v1.1.0/dokan_wix/WiXCA/CustomAction.cs). 

If so, can we rewrite the custom action in c/c++? I took a brief look at the custom action code and what it does seems to be invoking the dokanctl.exe program with proper arguments, which would register the dokan network provider in the registry (or remove it when uninstalling). So it should be not hard to rewrite it in c/c++.

Or do I miss something?

**Comment by Liryna** [comment]
Hi @lins05 ,

The vc redist is needed by the dokan library and driver not by the installer himself.
VC dependency is here to be sure that the computer has the environment to run dokan/seafile without issue.

If you are really looking to remove the VC dependency, it would need to build dokan with MT & create an installer without the dependency.
https://github.com/dokan-dev/dokany/issues/660

**Comment by lins05** [comment]
Thanks for the reply @Liryna ! 

I missed some info in the first post - actually the main pain point is the .NET framework, not the vc redist, that sometimes failed to install on old windows 7 systems. Do you think it's possible to remove the dep on the .NET framework if we rewrite the custom actions in c++? 

Of course if we can also remove the dep on vc redist (e.g by static linking as the linked issue you mentioned above), that would be an awesome bonus.

**Comment by Rondom** [comment]
This was discussed previously here:
https://github.com/dokan-dev/dokany/issues/254#issuecomment-225167493

It seems like it is not easily possible to port the code to .NET 3.5, which is included in Windows 7.

I see two options:
1. Update the installer to install not .NET 4.0, but 4.7.2 (latest version). Maybe this fixes the installation issues, because bugs have been fixed. This should of course only happen if .NET 4.0 or higher is not installed because that the requirement is still at least 4.0 and not 4.7.2.
2. Try to port the code.

I think pull requests are welcome for either of the two.

**Comment by Liryna** [comment]
Ok @lins05 I see. @Rondom is right. Normally we should detect and install the framework before running dokan installers.

@Rondom the 4.7.2 is not really the solution by looking at the version restricted windows version 😃  https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/versions-and-dependencies
4.6.1 is probably the best version to go if we would like to move.

But what am surprise is that 4.0 should be the best for Win 7, I really wonder why it would fail the installation. @lins05 do you have any error / log ? Maybe we can improve this part.

**Comment by lins05** [comment]
Thanks @Rondom @Liryna 👍 

>If we would like to rewrite in c++ the customaction. This seems to be possible:
https://github.com/rstropek/Samples/tree/master/WiXSamples/VisualStudioSnippetInstaller
Have to look if this does not create another issue where a redist need to be installed before..

IIUC a custom action is simply a function inside a DLL, and we even have a custom action written in plain C code: https://github.com/haiwen/seafile/blob/v6.1.7/msi/custom/custom.c#L71 . So I think there is nothing prevents us replacing the c# code with c/c++.

**Comment by Rondom** [comment]
> So I think there is nothing prevents us replacing the c# code with c/c++.

As much as nothing prevents you or someone else from porting it to .NET 3.5. For that it requires someone to spend the time to do the actual work and then do a lot of testing.... I think this is the main impedient.

**Comment by Rondom** [comment]
> Currently the code gives the same result as the C#. We are just missing the dokanctl output in case of error. This requires to redirect the output of CreateProcess well described here https://docs.microsoft.com/en-gb/windows/desktop/ProcThread/creating-a-child-process-with-redirected-input-and-output
I think I will open a ticket for this as evolution.

I think a first step would be printing the actual exit_code here: https://github.com/dokan-dev/dokany/commit/3812f49f1f73eb44f5d9d08cd38e5f14f1374baf#diff-63281fbea9278876f2794531d89624e8R42

Of course, the actual output would be more helpful to the user than an opaque error.

**Comment by Liryna** [comment]
@Rondom https://github.com/dokan-dev/dokany/commit/d725c58856a14ad58d025a815bdfef5c86daa808 done 👍 😃 

Humm I think even the output is mandatory here. dokanctl does not provide enough different return code. This should be done before next pre-release

**Comment by Liryna** [comment]
Dokanctl output redirection has been implemented https://github.com/dokan-dev/dokany/commit/5401ccb4ae11158fd387c6d17d91f08a823a2303
We now have the same behavior that before without the .NET dependency.
